http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/monitor/AutoDeleteAppMonitor.scala

```diff
@@ -0,0 +1,134 @@
+import org.broadinstitute.dsde.workbench.leonardo.{
+  AllowedChartName,
+  AppId,
+  AppName,
+  AppStatus,
+  CloudContext,
+  SamResourceId
+}
+import org.broadinstitute.dsde.workbench.model.{TraceId, WorkbenchEmail}
+import org.broadinstitute.dsde.workbench.openTelemetry.OpenTelemetryMetrics
+import org.broadinstitute.dsp.ChartName
+import org.typelevel.log4cats.StructuredLogger
+
+import java.time.Instant
+import scala.concurrent.ExecutionContext
+
+/**
+ * This monitor periodically sweeps the Leo database and auto delete App that have been running for too long.
+ */
+class AutoDeleteAppMonitor[F[_]](
+  config: AutoDeleteConfig,
+  publisherQueue: Queue[F, LeoPubsubMessage],
+  authProvider: LeoAuthProvider[F]
+)(implicit
+  dbRef: DbReference[F],
+  logger: StructuredLogger[F],
+  ec: ExecutionContext,
+  openTelemetry: OpenTelemetryMetrics[F]
+) extends BackgroundProcess[F, AppToAutoDelete] {
+  override def name: String =
+    "autoDeleteApp" //
+  override def interval: scala.concurrent.duration.FiniteDuration = config.autoDeleteCheckInterval
+
+  override def getCandidates(now: Instant)(implicit
+    F: Async[F],
+    metrics: OpenTelemetryMetrics[F],
+    logger: StructuredLogger[F]
+  ): F[Seq[AppToAutoDelete]] = for {
+    candidates <- appQuery.getAppsReadyToAutoDelete.transaction
+  } yield candidates
+
+  override def action(a: AppToAutoDelete, traceId: TraceId, now: Instant)(implicit F: Async[F]): F[Unit] =
+    for {
+      now <- F.realTimeInstant
+      loggingCtx = Map("traceId" -> traceId.asString)
+      _ <- a.cloudContext match {
+        case CloudContext.Gcp(googleProject) =>
+          if (a.appStatus == AppStatus.Error) {
+            implicit val implicitTraceId: Ask[F, TraceId] = Ask.const[F, TraceId](traceId)
+            for {
+              // we only need to delete Sam record for clusters in Google. Sam record for Azure is managed by WSM
+              _ <- authProvider.notifyResourceDeleted(
+                a.samResourceId,
+                a.creator,
+                googleProject
+              )
+              _ <- appQuery.markAsDeleted(a.id, now).transaction
+            } yield ()
+          } else {
+            for {
+              _ <- KubernetesServiceDbQueries.markPreDeleting(a.id).transaction
+              deleteMessage = DeleteAppMessage(
+                a.id,
+                a.appName,
+                googleProject,
+                None,
+                Some(traceId)
+              )
+              trackUsage = AllowedChartName.fromChartName(a.chartName).exists(_.trackUsage)
+              _ <- appUsageQuery.recordStop(a.id, now).whenA(trackUsage).recoverWith { case e: FailToRecordStoptime =>
+                logger.error(loggingCtx)(e.getMessage)
+              }
+              _ <- publisherQueue.offer(deleteMessage)
+            } yield ()
+          }
+        case CloudContext.Azure(_) => logger.info(loggingCtx)("Azure is not supported")
```

Comment:
Azure apps will have the threashold being 0, so they'll be filtered out already. The code as is will be easier to support Azure down the line as well

Reply:
- There is a log about "monitor starts". We will have that even without Azure apps
- When app is about to delete, there will be another log. Since Azure will not returned from getCandidates (because threshold is 0), there will be no log for Azure apps

Reply:
@mlilynolting  no...becuz they won't be returned to begin with from the query (filtered out in https://github.com/DataBiosphere/leonardo/pull/3943/files#diff-370c82ef7303ffa5487e72ee6d57f15a301110695a3a07f298f67a7f105190b1R379 already)...this branch here is not a normal path before we support Azure

Reply:
Oh, I see! Thanks 👍